Readme.md
# koa-http-stats

[![Build Status](https://travis-ci.org/vicanso/koa-http-stats.svg?style=flat-square)](https://travis-ci.org/vicanso/koa-http-stats)
[![Coverage Status](https://img.shields.io/coveralls/vicanso/koa-http-stats/master.svg?style=flat)](https://coveralls.io/r/vicanso/koa-http-stats?branch=master)
[![npm](http://img.shields.io/npm/v/koa-http-stats.svg?style=flat-square)](https://www.npmjs.org/package/koa-http-stats)
[![Github Releases](https://img.shields.io/npm/dm/koa-http-stats.svg?style=flat-square)](https://github.com/vicanso/koa-http-stats)

HTTP Stats middleware for Koa2. This is an useful simple stats, include: request total, request connecting total, request handle time and so on.

## Installation

```js
$ npm install koa-http-stats
```

## Examples

  View the [./examples](examples) directory for working examples.


### API

```js
const httpStats = require('koa-http-stats');
const Koa = require('koa');
const app = new Koa();


app.use(httpStats(function(performance, stats, ctx) {
	// {"total":1,"connecting":0,"status":[0,0,1,0,0,0],"time":[1,0,0,0,0,0],"size":[1,0,0,0,0,0],"busy":[1,0,0,0,0]}
	console.info(JSON.stringify(performance));
	// {"connecting":0,"total":1,"use":4,"bytes":11,"code":200,"status":2,"spdy":0,"size":0,"busy":0}
	console.info(JSON.stringify(stats));
}));

app.use(ctx => {
	if (ctx.url === '/wait') {
		return new Promise(function(resolve, reject) {
			ctx.body = 'Wait for 1000ms';
			setTimeout(resolve, 1000);
		});
	} else {
		ctx.body = 'Hello World';
	}
});

```

#### options

- `time` time stats config, default: `[30, 100, 500, 1000, 3000]`

- `size` size stats config, default: `[1024 * 2, 10 * 1024, 50 * 1024, 100 * 1024, 300 * 1024]`

- `status` status stats config, default: `[99, 199, 299, 399, 499]`

- `busy` connecting stats config, default: `busy: [50, 200, 500, 1000]`

#### onStats

When a request is done, the function will be trigger.

- `performance` application performance

- `stats` current http stats

- `ctx` koa.js context


## License

MIT


History.md
# 2.2.1
  * Parse the status code to integer

# 2.2.0
  * Remove `util` deprecated APIs


test/stats.js
'use strict';
const Koa = require('koa');
const httpStats = require('..');
const request = require('supertest');
const assert = require('assert');

describe('http-stats', done => {

	it('should set onStats success', done => {
		const app = new Koa();
		const onStats = (performance, statsResult) => {
			assert.equal(performance.total, 1);
			assert.equal(performance.connecting, 0);
			assert.equal(performance.status.length, 6);
			assert.equal(performance.time.length, 6);
			assert.equal(performance.size.length, 6);
			assert.equal(performance.status[2], 1);
			assert.equal(performance.time[0], 1);
			assert.equal(performance.size[0], 1);
			assert.equal(performance.busy[0], 1);

			assert(statsResult.use > 0);
			assert.equal(statsResult.bytes, 11);
			assert.equal(statsResult.status, 2);
			assert.equal(statsResult.spdy, 0);
			assert.equal(statsResult.size, 0);
			assert.equal(statsResult.busy, 0);
		};
		app.use(httpStats(onStats));
		app.use(ctx => {
			if (ctx.url === '/wait') {
				return new Promise(function(resolve, reject) {
					ctx.body = 'Wait for 1000ms';
					setTimeout(resolve, 1000);
				});
			} else {
				ctx.body = 'Hello World';
			}
		});

		const server = app.listen();
		request(server)
			.get('/')
			.expect(200, 'Hello World')
			.end(function(err, res) {
				done(err);
			});

	});

	it('should get stats success when an error throw', done => {
		const app = new Koa();
		const onStats = (performance, statsResult) => {

			assert.equal(statsResult.code, 500);
			assert.equal(statsResult.status, 5);
			assert(statsResult.use > 1000);
			assert.equal(statsResult.spdy, 4);
		};
		app.use(httpStats(onStats));
		app.use(ctx => {
			return new Promise((resolve, reject) => {
				setTimeout(() => {
					const err = new Error('ERROR');
					err.status = 'ECONNABORTED';
					reject(err);
				}, 1000);
			});
		});

		request(app.listen())
			.get('/')
			.expect(500, done);
	});

	it('should set options success', done => {
		const app = new Koa();
		const onStats = (performance, statsResult) => {
			assert.equal(statsResult.status, 3);
		};
		app.use(httpStats({
			status: [1, 2, 3]
		}, onStats));
		app.use(ctx => {
			ctx.body = 'Hello World';
		});

		request(app.listen())
			.get('/')
			.expect(200, done);
	});


	it('should get connecting count success', done => {
		let hasChecked = false;
		const app = new Koa();
		const onStats = (performance, statsResult) => {
			if (!hasChecked) {
				assert.equal(performance.connecting, 1);
				hasChecked = true;
			}
		};
		app.use(httpStats(onStats));
		app.use(ctx => {
			if (ctx.url === '/wait') {
				return new Promise(function(resolve, reject) {
					ctx.body = 'Wait for 1000ms';
					setTimeout(resolve, 1000);
				});
			} else {
				ctx.body = 'Hello World';
			}
		});

		const server = app.listen();

		request(server)
			.get('/wait')
			.expect(200, 'Wait for 1000ms')
			.end(done);
		request(server)
			.get('/')
			.expect(200, 'Hello World')
			.end((err) => {
			});
	});
});

index.js
'use strict';

function noop() {}

function isFunction(fn) {
  return typeof fn === 'function';
}
function isObject(x) {
  return x != null && typeof x === 'object';
}

function fill(total, v) {
  const arr = [];
  for (let i = 0; i < total; i += 1) {
    arr.push(v);
  }
  return arr;
}

function extendOptions(options) {
  const opts = options;
  const defaultOptions = {
    time: [30, 100, 500, 1000, 3000],
    size: [1024 * 2, 10 * 1024, 50 * 1024, 100 * 1024, 300 * 1024],
    status: [99, 199, 299, 399, 499],
    busy: [50, 200, 500, 1000],
  };
  Object.keys(defaultOptions).forEach((key) => {
    /* istanbul ignore else */
    if (!opts[key]) {
      opts[key] = defaultOptions[key];
    }
  });
}

function sortedIndex(arr, value) {
  let low = 0;
  let high = arr.length;
  while (low < high) {
    /* eslint no-bitwise:0 */
    const mid = (low + high) >>> 1;
    const computed = arr[mid];
    if (computed < value) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }
  return high;
}

function get(arr, filter, defaultValue) {
  let result;
  arr.forEach((tmp) => {
    if (tmp && filter(tmp)) {
      result = tmp;
    }
  });
  return result || defaultValue;
}

function stats() {
  const args = Array.from(arguments);
  const onStats = get(args, isFunction, noop);
  const options = get(args, isObject, {});
  extendOptions(options);
  const performance = {
    total: 0,
    connecting: 0,
    status: fill(options.status.length + 1, 0),
    time: fill(options.time.length + 1, 0),
    size: fill(options.size.length + 1, 0),
    busy: fill(options.busy.length + 1, 0),
  };

  return (ctx, next) => {
    const start = Date.now();
    performance.total += 1;
    performance.connecting += 1;


    function done(err) {
      const use = Date.now() - start;
      performance.connecting -= 1;
      const connecting = performance.connecting;

      let statusCode = ctx.status;
      /* istanbul ignore else */
      if (err) {
        statusCode = parseInt(err.status || err.statusCode, 10);
        if (Number.isNaN(statusCode)) {
          statusCode = 500;
        }
      }

      const bytes = ctx.length || 0;
      const statsResult = {
        connecting,
        total: performance.total,
        use,
        bytes,
        code: statusCode,
      };
      let index = sortedIndex(options.status, statusCode);
      statsResult.status = index;
      performance.status[index] += 1;

      index = sortedIndex(options.time, use);
      statsResult.spdy = index;
      performance.time[index] += 1;

      index = sortedIndex(options.size, bytes);
      statsResult.size = index;
      performance.size[index] += 1;

      index = sortedIndex(options.busy, connecting);
      statsResult.busy = index;
      performance.busy[index] += 1;

      onStats(performance, statsResult, ctx);
      if (err) {
        throw err;
      }
    }
    return next().then(done, done);
  };
}


module.exports = stats;
